### ISSUE TITLE
Issues related to numpy

### ISSUE THREAD

**Issue by hauntsaninja** [python/mypy#18343]
This can be a tracking issue.

See also https://github.com/numpy/numpy/issues/27957

**Comment by hauntsaninja** [comment]
PR adding mypy_primer to numpy's CI: https://github.com/numpy/numpy/pull/28073

**Comment by hauntsaninja** [comment]
PR fixing a crash when using numpy with weird import following settings: https://github.com/python/mypy/pull/18351

**Comment by hauntsaninja** [comment]
PR improving diagnostics for some case someone ran into using numpy: https://github.com/python/mypy/issues/16875

**Comment by jorenham** [comment]
Not sure if this has been reported before, but the following line is causing stubtest to crash with `numpy>=2.2.0` in case you have an empty numpy array as argument default (which is what least `scipy==1.14.1` does):

https://github.com/python/mypy/blob/60da03aae5cade43e106f6e33fcd26f524e63213/mypy/stubtest.py#L673

Because since NumPy 2.2.0, `ndarray.__bool__` will raise if the array is empty. And the reason why `__bool__` gets called, is because `ndarray.__ne__` returns another array with `np.bool` dtype of the same shape:

```pycon
>>> import numpy as np
>>> np.__version__
'2.2.0'
>>> a = np.array([])
>>> a
array([], dtype=float64)
>>> a == object()  # numpy "vectorizes" the `==` over each element
array([], dtype=bool)
>>> bool(a == object())
Traceback (most recent call last):
  File "<python-input-24>", line 1, in <module>
    bool(a == object())
    ~~~~^^^^^^^^^^^^^^^
ValueError: The truth value of an empty array is ambiguous. Use `array.size > 0` to check that an array is not empty.
```

So when stubtest encounters a `def f(a=np.array([])): ...` (on the `.py` side), `stubtest` will crash:

```shell
$ stubtest --mypy-config-file=pyproject.toml --allowlist=.mypyignore --ignore-unused-allowlist scipy
Traceback (most recent call last):
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/bin/stubtest", line [8](https://github.com/jorenham/scipy-stubs/actions/runs/12246812991/job/34163454128?pr=288#step:7:9), in <module>
    sys.exit(main())
             ~~~~^^
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line 20[9](https://github.com/jorenham/scipy-stubs/actions/runs/12246812991/job/34163454128?pr=288#step:7:10)8, in main
    return test_stubs(parse_options(sys.argv[1:]))
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line 1971, in test_stubs
    for error in test_module(module):
                 ~~~~~~~~~~~^^^^^^^^
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line 247, in test_module
    yield from verify(stub, runtime, [module_name])
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line 423, in verify_mypyfile
    yield from verify(stub_entry, runtime_entry, object_path + [entry])
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line 578, in verify_typeinfo
    yield from verify(stub_to_verify, runtime_attr, object_path + [entry])
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line [10](https://github.com/jorenham/scipy-stubs/actions/runs/12246812991/job/34163454128?pr=288#step:7:11)64, in verify_funcitem
    for message in _verify_signature(stub_sig, runtime_sig, function_name=stub.name):
                   ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.[13](https://github.com/jorenham/scipy-stubs/actions/runs/12246812991/job/34163454128?pr=288#step:7:14)/site-packages/mypy/stubtest.py", line 902, in _verify_signature
    yield from _verify_arg_default_value(stub_arg, runtime_arg)
  File "/home/runner/work/scipy-stubs/scipy-stubs/.venv/lib/python3.13/site-packages/mypy/stubtest.py", line 664, in _verify_arg_default_value
    if runtime_arg.default != inspect.Parameter.empty:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: The truth value of an empty array is ambiguous. Use `array.size > 0` to check that an array is not empty.
Error: Process completed with exit code 1.
```

Luckily, fixing this is just a matter of replacing `!=` with `is not` when comparing against `inspect.Parameter.empty` (which happens in at least 2 two other places in `mypy/stubtest.py` as well, btw).

I could scrape together a PR if you'd like, but that could take some time (of which I don't have a lot the coming week) as I don't have a local dev env for mypy setup at the moment. And I can imagine that's more mypy-savvy could probably fix this in a coffee break on their nokia 3310 (assuming they don't get distracted and end up playing snake 2, of course). Anyway, I wouldn't mind either way, so let me know 🤷🏻.

**Comment by hamdanal** [comment]
> Not sure if this has been reported before, but the following line is causing stubtest to crash with `numpy>=2.2.0` in case you have an empty numpy array as argument default (which is what least `scipy==1.14.1` does):

Thanks for the report, fix is here #18353

**Comment by jorenham** [comment]
https://github.com/python/mypy/issues/18359 and https://github.com/python/mypy/issues/18257 also have a big impact on numpy, which I explained in https://github.com/python/mypy/issues/18359

**Comment by jorenham** [comment]
It's not possible to annotate the following behavior:

```pycon
>>> class Spam: ...
>>> np.object_(Spam())
<__main__.Spam object at 0x7f536d7b5160>
```

And it leads to incorrectly inferred types

```pyi
class Spam: ...
reveal_type(np.object_(Spam()))   # numpy.object_  (wrong; should be `Spam`)
```

<sub>Pyright correctly reports `Type of "np.object_(Spam())" is "Spam"`.</sub>

---

The other numpy "scalar" constructors either return an instance of themselves, or an `ndarray`, dependening on the input. For example:

```pycon
>>> np.int8(42)
np.int8(42)
>>> np.int8([42])
array([42], dtype=int8)
```

The bare minimum stub looks like 

```pyi
class int8(float):
    @overload
    def __new__(cls, x: int = 0, /) -> Self: ...
    @overload
    def __new__(cls, x: Sequence[int], /) -> np.ndarray[tuple[int], np.dtype[np.int8]]: ...
```

But a false positive (according to the [typing spec](https://typing.readthedocs.io/en/latest/spec/constructors.html#new-method)) is reported:

```
Incompatible return type for "__new__" (returns "ndarray[tuple[int], dtype[signedinteger[_8Bit]]]", but must return a subtype of "int8")
```

But even so, the second overload isn't completely ignored, because the following isn't reported as an error:

```pyi
reveal_type(int8(42))  # int8  (correct)
reveal_type(int8([42]))  # int8  (wrong; should be `numpy.ndarray`)
```

So it's as if the return type of second `__new__` overload is ignored, and `Self` is used instead 🤔

---

I believe that https://github.com/python/mypy/issues/15182 is at least in part responsible for this behavior.

---

Oh and you might already know this, but this is also blocking a typeshed fix for (at least) `builtins.reversed` annotations (https://github.com/python/typeshed/pull/11646). So the amount of people that will be happy once this is resolved isn't limited to the numpy community :)

**Comment by jorenham** [comment]
https://github.com/python/mypy/issues/18496

**Comment by matteobrv** [comment]
https://github.com/python/mypy/issues/18566

**Comment by JukkaL** [comment]
More lenient checking of variadic generics (needs discussion):
* #18665

**Comment by JukkaL** [comment]
I'm working on making variable redefinitions more flexible, which should help NumPy:
* #6232
* #6233

**Comment by jorenham** [comment]
There's also another stubtest issue that's affecting nummy, this time related to a positional-or-keyword argument prefixed with `__`:
https://github.com/python/mypy/issues/15302#issuecomment-2701979355

**Comment by jorenham** [comment]
> I'm working on making variable redefinitions more flexible, which should help NumPy:
> 
> * [Allow variable redefinition if lifetimes don't overlap #6232](https://github.com/python/mypy/issues/6232)
> * [Infer union from conditional definition in if statement #6233](https://github.com/python/mypy/issues/6233)

The relevant PR: https://github.com/python/mypy/pull/18727

**Comment by hauntsaninja** [comment]
PR fixing the `__` prefix thing https://github.com/python/mypy/pull/18756

**Comment by hauntsaninja** [comment]
Crash on numtype with pre-release mypy https://github.com/python/mypy/issues/18764

**Comment by jorenham** [comment]
This is a big one: #18997

**Comment by jorenham** [comment]
As a consequence of https://github.com/python/mypy/issues/19341, mypy will incorrectly infer `[1, 2] * np.array([3, 4], np.float64)` as `list[int]`, even though `ndarray.__index__` is annotated with `self: NDArray[integer]` where `integer` is incompatible with `float64`. 

For those unfamiliar with NumPy this could look pretty strange, but In NumPy it's pretty common to mix "array-likes" (e.g. a `list[int]`) and native numpy `ndarray` types (e.g. returned by  `np.array()`). It's very convenient for users, but can be pretty horrible to annotate haha.

**Comment by ilevkivskyi** [comment]
@jorenham Just to give some large scale update here: most-likely the mypy core team will be busy with various performance-related work until the end of the year. But hopefully I will have time to make some dedicated effort towards better numeric/scientific stack support ~January 2026 (or maybe earlier). However, if there are any pressing bug(s) that you think we should prioritize, please let us know. Maybe I can spend some time (especially if there is a simple fix).

Btw I noticed mypy plugin is deprecated https://github.com/numpy/numpy/blob/6bb72644e16e5e7f590533bde670c42b199edb53/numpy/typing/mypy_plugin.py#L20, why is it deprecated?

IMO this is a bad decision, a lot of things in Python cannot be type-checked properly (even in theory), so plugins can often be very helpful.

**Comment by jorenham** [comment]
> But hopefully I will have time to make some dedicated effort towards better numeric/scientific stack support ~January 2026 (or maybe earlier).

That's awesome!

> However, if there are any pressing bug(s) that you think we should prioritize, please let us know. Maybe I can spend some time (especially if there is a simple fix).

At the moment there are no high-prio issues that come to mind. There are some issues that I keep running into, but those are all rather involved, I believe. But I'll maybe there's some I'm forgetting right now, in which case I'll be sure to let you know. Thanks!

> Btw I noticed mypy plugin is deprecated [numpy/numpy@`6bb7264`/numpy/typing/mypy_plugin.py#L20](https://github.com/numpy/numpy/blob/6bb72644e16e5e7f590533bde670c42b199edb53/numpy/typing/mypy_plugin.py#L20), why is it deprecated?

The only thing the plugin was used for, was for "specializing" platform dependent scalar types. For example, it would effectively replace `np.intp` with `np.int32` on 32-bit platforms, and `np.int64` on 64-bit platforms. 
Technically, it's more accurate this way, but in practice there were barely any use-cases where it actually helped. In fact, I haven't seen any examples of projects where disabling the plugin caused any issue or different type-checking behavior.

> IMO this is a bad decision, a lot of things in Python cannot be type-checked properly (even in theory), so plugins can often be very helpful.

That's certainly true. And if you solely support mypy, then that's no problem. NumPy, however, aims to support all typing-spec-compliant type-checkers. So we try to minimize the differences in the behavior of type-checker. But by using a mypy plugin, those differences increase.

But if there are any projects that *rely* on NumPy's mypy plugin, and where removing it would cause problems, then I'd be open to talking about reconsidering the deprecation.

**Comment by ilevkivskyi** [comment]
> That's certainly true. And if you solely support mypy, then that's no problem. NumPy, however, aims to support all typing-spec-compliant type-checkers. So we try to minimize the differences in the behavior of type-checker. But by using a mypy plugin, those differences increase.

I don't follow the logic here: how writing a mypy plugin will make types worse for other type checkers?

**Comment by jorenham** [comment]
> how writing a mypy plugin will make types worse for other type checkers?

It wouldn't make it worse for other type-checkers, it would increase the differences between them.

**Comment by ilevkivskyi** [comment]
I mean why this is bad? Like no one is allowed to have better type-checking unless everyone can have it? Anyway, I guess someone else will write it (maybe even us).

**Comment by jorenham** [comment]
> I mean why this is bad? Like no one is allowed to have better type-checking unless everyone can have it?

In this case it's bad because it causes mypy (and only mypy) to infer different types on different platforms, which is bound to lead to issues of the "but it works on my machine" kind. 

Another reason is that I'd like to avoid having to ask "what type checker are you using?" when someone opens a typing-related issue. 

And when working on the stubs, ideally you wouldn't have to fine-tune everything for specific type-checkers, since that requires quite a lot of mental effort to track, and in practice tends to lead to comments like `# note: mypy treats this as ...`.
It's *easier* to think of "type-checkers" as a whole, instead of having to think of each individual type-checkers.

**Comment by jorenham** [comment]
I just opened https://github.com/python/mypy/issues/20223; a stubtest issue related to numpy ufunc's, and a proposed solution.

**Comment by tyralla** [comment]
@jorenham: We now in fact consider including an (optional) NumPy plugin in Mypy.  Personally, I am longing for fewer unexpected `Any` return values and better shape typing (possibly including simple math).  See [this issue](https://github.com/hydpy-dev/hydpy/issues/181) for a (very rapidly prototyped) prototype. 

I know you prefer some degree of synchronicity among type checkers.  But are you nevertheless interested in sharing your opinion on which plugin features could help a significant number of NumPy and Mypy users?

**Comment by jorenham** [comment]
> [@jorenham](https://github.com/jorenham): We now in fact consider including an (optional) NumPy plugin in Mypy. Personally, I am longing for fewer unexpected `Any` return values and better shape typing (possibly including simple math). See [this issue](https://github.com/hydpy-dev/hydpy/issues/181) for a (very rapidly prototyped) prototype.
> 
> I know you prefer some degree of synchronicity among type checkers. But are you nevertheless interested in sharing your opinion on which plugin features could help a significant number of NumPy and Mypy users?

I noticed that you use `Literal` integers in the shape type in your example. However, as I explained in https://discuss.python.org/t/pep-827-type-manipulation/106353/88 and https://github.com/numpy/numtype/issues/578, I would recommend against doing that. 

The `cumsum` example in the referenced issue is something that could (and should) be fixed in the numpy stubs. It should be too difficult, since it's a shape-preserving operation. So if you truly wish to help a significant number of NumPy users, then I'd say contributing stubs improvements to this class of functions would be the way to go.

However, there some cases where it's currently impossible to statically determine the output shape type, but where a mypy plugin *would* be able to. An rather extreme example of this is [`numpy.einsum`](https://numpy.org/doc/stable/reference/generated/numpy.einsum.html), where the output shape is determined through a string using a particular mini-syntax. 

Then there's also a third category of functions where even a mypy plugin wouldn't be able to determine the shape-type, for example [`numpy.load`](https://numpy.org/doc/stable/reference/generated/numpy.load.html#numpy.load).

So to summarize; a mypy plugin for shape-typing would be helpful for functions whose output shape-type can statically be determined yet cannot be expressed using Python's type system. I'm not sure which ones exactly these are, but there are probably only a handful of them.

**Comment by tyralla** [comment]
Thanks for your insights, @jorenham!

> I noticed that you use `Literal` integers in the shape type in your example. However, as I explained in https://discuss.python.org/t/pep-827-type-manipulation/106353/88 and [numpy/numtype#578](https://github.com/numpy/numtype/issues/578), I would recommend against doing that.

I know you thought about this much more thoroughly, but I am still optimistic that shape typing (using `Literal`) would be very helpful and may not be impossible (though maybe too much effort).

As I see it, the main points in your cited comments are...

**Upcasting problems (`Literal[1]` to `int`)**

Opposed to basedpyright, Mypy keeps the `Literal` type in the given example:

```python
from typing import Literal

def unwrap[N: int](shape: tuple[N]) -> N:
    return shape[0]

s: tuple[Literal[1]] = (1,)
reveal_type(unwrap(s))  # note: Revealed type is "Literal[1]"
```

**Missing integer arithmetic logic in type checkers**

That would be one task of the pugin.  We could teach it that `np.reshape(np.ndarray[tuple[Literal[12]], Any], (4, -1)) -> np.ndarray[tuple[Literal[4], Literal[3]], Any]`. (Admittedly, lots of work.)

**False positives (or, more correctly, unclear positives)**

If a function only works for `ndarray[tuple[Literal[2], Literal[2]], Any]`, I would be happy if Mypy would complain if I insert `ndarray[tuple[int, int], Any]`


> The `cumsum` example in the referenced issue is something that could (and should) be fixed in the numpy stubs. It should be too difficult, since it's a shape-preserving operation. So if you truly wish to help a significant number of NumPy users, then I'd say contributing stubs improvements to this class of functions would be the way to go.

I picked this example absolutely randomly.  But yes, making the stubs perfect where possible is preferable, of course.  For me, it is hard to predict which parts of the NumPy stubs can be (or will be) improved substantially.

Thanks for your recommendations again. I'll let them sink in for a while.

**Comment by ilevkivskyi** [comment]
As someone who used to be an active NumPy user for ~10 years (a little known fact about me: I have PhD in theoretical physics) here are my 2 cents:
* There are two different levels of "precision" when typing arrays: just the number of dimensions, and actual array shape. In my experience it would be great to have the second. There were countless times when after waiting 5 minutes I discovered that I have an off-by-one error because of `np.diff()`, or some matrices won't multiply because I forgot `.T`.
* I don't really buy the argument that (exact) `Literal` types for array shapes will cause a lot of harm because we don't have a "gradual" equivalent of `int`. First, if plugin would only (or mostly) infer precise types for arrays (not for functions that accept them), this will not be a problem unless an array appears in invariant position, which is rarely the case. Second, in the worst case we can always add `mypy_extensions.AnyInt`.

**Comment by jorenham** [comment]
> * I don't really buy the argument that (exact) `Literal` types for array shapes will cause a lot of harm because we don't have a "gradual" equivalent of `int`. First, if plugin would only (or mostly) infer precise types for arrays (not for functions that accept them), this will not be a problem unless an array appears in invariant position, which is rarely the case. Second, in the worst case we can always add `mypy_extensions.AnyInt`.

If such an `AnyInt` would exist in Python, and if there'd be some way to statically infer that e.g. `np.empty(12).reshape((3,-1))` has shape `(3, 4)`, that `np.swapaxes([[1, 2, 3], [4, 5, 6]], -1, -2)` has shape `(3, 2)`, etc., then I'd be all for it. 

But don't get me wrong; AFAIK it perfectly fine if someone creates a mypy plugin to help with shape-typing in NumPy. It's just not something that we're going to build ourselves or require our users to use; that's all.

**Comment by ilevkivskyi** [comment]
@jorenham @MarcoGorelli Could you please add more links to issues affecting Python numeric stack here?

**Comment by jorenham** [comment]
There's https://github.com/python/mypy/issues/11347 amd https://github.com/python/mypy/issues/14070 that can lead to unexpected `np.ndarray` inferred dtypes in certain situations. We often use the "Never trick" to work around this (because pyright also has a similar issue), but there are still some places where it can still happen.

There are also quite a lot of `# type: ignore[override]` in the numpy stubs that I believe to be false positives. And although I haven't investigated this in detail, I wouldn't be surprised if at least some of those are related to https://github.com/python/mypy/issues/12379 and https://github.com/python/mypy/issues/20557 .

I also ran into https://github.com/python/mypy/issues/9464 a couple of times. But Eric Traut once told me that fixing this could be very complex, because it would require solving constraint sets using something like a unification algorithm. But if mypy already has the machinery for that in place, perhaps it's not as difficult as it would be for pyright 🤷

**Comment by jorenham** [comment]
Mypy crashes in mypy_primer on the latest numpy branch: https://github.com/python/mypy/issues/21824

It's a pretty interesting one, because it depends on how you *name* a variable.
